src/client/dfuse/dfuse.h

```diff
@@ -138,6 +138,75 @@ struct dfuse_obj_hdl {
 	bool                      doh_kreaddir_finished;
 };
 
+/* Readdir support.
+ *
+ * Readdir is by far the most complicated component of dfuse as a result of the kernel interfaces,
+ * the dfs interface and the lack of kernel caching for concurrent operations.
+ *
+ * The kernel interface to readdir is to make a callback into dfuse to request a number of dentries
+ * which are then populated in a buffer which is returned to the kernel, each entry in the buffer
+ * contains the name of the entry, the type and some other metadata.  The kernel requests a buffer
+ * size and the dfuse can reply with less if it choses - 0 is taken as end-of-directory.  The
+ * length of the filenames affects the number of entries that will fit in the buffer.
```

Comment:
> the dfuse can reply with less if it choses

choses -> chooses
But does it choose to? If so, when?

> The length of the filenames affects the number of entries that will fit in the buffer.

maybe be more direct? Something like:
"The buffer is populated with as many filenames as fit."
or "The filenames are stored sequentially in the buffer, up to the buffer size."

Reply:
I don't think that's true.  If the readdir call had to fetch more data from dfs then it'll call iterate but if it runs our again then it might return with a partially empty buffer.  You will also probable have a partially empty buffer on end-of-directory.

Reply:
that's an internal libfuse detail.